### PR TITLE
Fix issue where root.os.panic could return

### PR DESCRIPTION
**Pull request by rgreenblatt**
This change forces root.os.panic to have the `PanicFn` type. I think this also simplifies the code somewhat.

This could break existing code somewhere.

Edit with more detail on the issue this PR solves:
The issue is that `root.os.panic` was used without checking the type of the function (specifically, without checking the type is `PanicFn`). `PanicFn` has return type `noreturn`, but because `root.os.panic` isn't checked, the function could have return type `void`. For this reason, there was `unreachable` after `root.os.panic`. So if you accidently wrote a panic handler which returned, then you would get issues at runtime (hitting the unreachable and causing another panic or undefined behavior) instead of a compiler error. 

Additionally, the code for using the override `root.os.panic` was different than the code for using the override `root.panic` for no good reason.

### CODE REVIEW

File: lib/std/builtin.zig [hunk 1]
```diff
@@ -664,16 +664,17 @@ pub const PanicFn = fn ([]const u8, ?*StackTrace) noreturn;
...
+    root.os.panic
+else
+    default_panic;
 
 /// This function is used by the Zig language code generation and
 /// therefore must be kept in sync with the compiler implementation.
 pub fn default_panic(msg: []const u8, error_return_trace: ?*StackTrace) noreturn {
     @setCold(true);
-    if (@hasDecl(root, "os") and @hasDecl(root.os, "panic")) {
-        root.os.panic(msg, error_return_trace);
```
Comment (matu3ba) [inline lib/std/builtin.zig:674]: Why did you remove the error return trace here?

File: lib/std/builtin.zig [hunk 2]
```diff
@@ -664,16 +664,17 @@ pub const PanicFn = fn ([]const u8, ?*StackTrace) noreturn;
 
 /// This function is used by the Zig language code generation and
 /// therefore must be kept in sync with the compiler implementation.
-pub const panic: PanicFn = if (@hasDecl(root, "panic")) root.panic else default_panic;
+pub const panic: PanicFn = if (@hasDecl(root, "panic"))
+    root.panic
+else if (@hasDecl(root, "os") and @hasDecl(root.os, "panic"))
+    root.os.panic
```
Comment (matu3ba) [inline lib/std/builtin.zig:670]: This case has no cold cache hint in contrast to the original code.

**Review by matu3ba — CHANGES_REQUESTED** [review]
Please link the related issue in this PR with a proper description.

File: lib/std/builtin.zig (hunk 2 above)
Comment (rgreenblatt) [reply lib/std/builtin.zig:670]: I don't understand. `@setCold(true)` is still the first line in `default_panic`.

Perhaps you aren't recognizing that this if statement is comptime? Or maybe I am totally missing your point?

File: lib/std/builtin.zig (hunk 1 above)
Comment (rgreenblatt) [reply lib/std/builtin.zig:674]: This code isn't needed because default_panic is no longer used when `root.os.panic` is defined.

**Comment by rgreenblatt** [comment]
> Please link the related issue in this PR with a proper description.

This is a (small) subset of https://github.com/ziglang/zig/pull/8228 which I have extracted out. There isn't a corresponding issue.

The aim is just to properly typecheck `root.os.panic`.

File: lib/std/builtin.zig (hunk 2 above)
Comment (matu3ba) [reply lib/std/builtin.zig:670]: my bad, you are correct here.

File: lib/std/builtin.zig (hunk 1 above)
Comment (matu3ba) [reply lib/std/builtin.zig:674]: Currently the language reference does not mention os-specific panic handler or what the expected behavior is, so I feel unsure about this. 
I would suggest changing doc.langref.html.in starting at `{#header_open|@panic#}` to force clarification on this or ping/talk to somebody who knows how it stuff was agreed to work and quote it.

**Comment by matu3ba** [comment]
Sorry to be picky, but ideally you would give a case where this can happen. Maybe a combination of custom and/or default panic handler?

**Comment by rgreenblatt** [comment]
> Sorry to be picky, but ideally you would give a case where this can happen. Maybe a combination of custom and/or default panic handler?

The issue is that `root.os.panic` was used without checking the type of the function (specifically, without checking the type is `PanicFn`). `PanicFn` has return type `noreturn`, but because `root.os.panic` isn't checked, the function could have return type `void`. For this reason, there was `unreachable` after `root.os.panic`. So if you accidently wrote a panic handler which returned, then you would get issues at runtime instead of a compiler error. 

Additionally, the code for using the override `root.os.panic` was different than the code for using the override `root.panic` for no good reason.

This PR addresses both of these issues. Sorry about not clarifying exactly what this PR does and why.

File: lib/std/builtin.zig (hunk 1 above)
Comment (rgreenblatt) [reply lib/std/builtin.zig:674]: Yeah unfortunately all of this is undocumented. In theory I think the docs would go here: https://ziglang.org/documentation/master/#Root-Source-File
